### ISSUE TITLE
Operation encoding

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#203]
We should start thinking about how each operation should be encoded (right now we basically assign dummy opcodes to operations). There are many things to consider, but for now, I'll just describe the basics.

First, we want to encode each operation with no more than 7 bits. This also means that within the VM we allocate 7 columns for operation bits (or `op_bits`). Theoretically, we could encode up to 128 different operations using seven bits - but this is not desirable for a number of reasons:

1. 128 is a large number and we'd like to keep the number of operations as small as possible (while not giving up expressivity).
2. With 128 operations, each operation flag would have degree 7. While this should work for most operations, it won't work for all (e.g., many control flow operations should have degree 5, or we'll need to introduce extra columns for degree reduction).

If we keep the number of operations to at most 96, we can create several classes of operations. Consider the following arrangement:

| bit 6 | bit 5 | bit 4 | bit 3 | bit 2 | bit 1 | bit 0 | # of ops | degree |
| :---: | :---: | :---: | :---: | :----: | :---: | :---: | :-------: | :------: |
| 0     |  x      | x      | x       | x       | x       | x      |  64          |  7         |
| 1     |  0      |  0     | x       | x       | x       | x      |  16          |  7         |
| 1     |  0      |  1     | x       | x       | x       | -       |  8           |  6         |
| 1     |  1      |  x     | x       | x       | -       | -       |  8           |  5         |

In the above, any opcode which starts with `0` can use all of the following 6 bits and thus we get 64 total opcodes and degree of each opcode is 7. Opcodes which start with `100` work in the same way.

However, for opcodes which start with `101` we allow using only 6 bits total. Thus, we can encode 8 opcodes, and degree of each opcode will be 6. We apply the same logic to opcodes which start with `11` but for these opcodes we allow using only 5 bits total. We still can encode 8 opcodes, but the degree of each of these opcodes is now 5.

To summarize, with this arrangement, we get 3 classes of opcodes:

- 80 opcodes of degree 7
- 8 opcodes of degree 6
- 8 opcodes of degree 5

**Comment by bobbinth** [comment]
Another thing to keep in mind is how many 16-bit range checks a given operation requires. Here is a first stub at operations which require range checks.

| Operation   | # of checks |
| ----------- | ------------ |
|  `U32split` | 4                  |
| `U32add`   | 2                 |
| `U32addc` | 2                 |
| `U32sub`  | 2                 |
| `U32mul`   | 4                 |
| `U32madd`    | 4             |
| `U32div`         | 4             |
| `U32assert2` | 4             |

Degree of the selectors for operations which require 2 range checks should not exceed 6. For the ones which require 4 range checks, it should not exceed 4.

For the ones which require 2 range checks, we could create a common prefix. Something like `10011`. This prefix would have degree 5, and would allow us to have 4 operations requiring exactly 2 range checks. We currently have 3 - so, we could add one more such operation.

For the ones requiring 4 range checks, the things are a bit more problematic. We could also define a common prefix, something like `1010`, and it would have degree 4, but this would accomodate only 4 out of 5 required operations. So, what to do with the 5th one still needs to be decided.

**Comment by bobbinth** [comment]
One other thought: if we change `U32addc` to something like `U32add3` (i.e., add top 3 items on the stack), this operation would also require 4 range-checks. So, we'll have 6 operations which require 4 range checks. We can then force `U32add` and `U32sub` to perform two 2 extra (unneeded) range checks, and we'll have 8 operations which require 4 range checks each.


This would result in a bit of waste, but:
1. Would make the constraints a bit more uniform (which would compensate somewhat for the extra waste).
2. We could group these 8 operations under a common prefix (e.g., `1000`) and the degree of this prefix would be 4. So, no extra thinking is needed.

**Comment by grjte** [comment]
A few thoughts on operation encodings and flags:

1. There are currently 17 operations which shift the stack left, 2 of which require range checks (`u32add3`, `u32madd`). The operations not requiring range checks are  `add`, `mul`, `and`, `or`, `eq`, `loadw`, `storew`, `drop`, `cswap`, `cswapw`, `assert`, `fmpupdate`, `u32and`, `u32or`, `u32xor`.
2. There are currently 18 operations which shift the stack right, 1 of which requires range checks (`u32split`). The operations not requiring range checks are `eqw`, `push`, `read`, `sdepth`, `pad`, `dup0`-`dup7`, `dup9`, `dup11`, `dup13`, `dup15`.

## Operations that shift the stack left

I haven't come up with a way to group all left shifts and all right shifts under a single flag while also keeping the flag that specifies range checks (prefix 0b1000). For left shifts, there are 15 operations which do not require range checks, so if we waste one opcode then we can group them all under a shared flag of degree 3. This would result in 2 flags that indicate left shifts, one of which is specific to the operations that also use range checks (e.g. 100010x).


## Operations that shift the stack right

For right shifts, things are trickier, because there are 17 right shift operations that do not require range checks, so we either need 3 different flags (1 of degree 3 specifying 16 operations, 1 that specifies the range check operation, and one that specifies the remaining operation), or we may want to consider removing one of the right shift operations (e.g. one of the `dup` operations) or changing it so it no longer shifts the stack. 

For example, `sdepth` is used infrequently , so if the processor operation overwrote the top value of the stack (instead of pushing) and parser transformed this operation into `Pad SDepth` the only stdlib function it would currently affect is `finalize_stack`. This would allow us to group all remaining right shift operations under a single flag. Additionally, it's possible that we would want any future environment operations to be implemented somewhat analogously to `sdepth`. If they were also to be implemented as operations that shift the stack, this would cause the right shift flag to be affected in the future.

**Comment by bobbinth** [comment]
> There are currently 17 operations which shift the stack left

I think there are actually more. Some of the control flow operations also shift the stack to the left. Specifically, `SPLIT`, `LOOP` (but only if the top of the stack is `1`), `REPEAT`, and `END` (but only when exiting a loop block). I don't think this presents an issue for grouping though, as control flow operations would be grouped separately anyway.

> For left shifts, there are 15 operations which do not require range checks, so if we waste one opcode then we can group them all under a shared flag

I think this is fine. I'm sure we'll come up with some operation later that will require a left shift.

> For right shifts, things are trickier, because there are 17 right shift operations that do not require range checks, so we either need 3 different flags

I think this if fine for now - 3 flags is not too bad. I do like the idea of changing `sdepth` if needed, but we can do it later. 

There is one other consideration regarding left and right shifts: it might make sense to define multiple shift flags for each type of shift. This may help with constraint definitions later.

For example, for right shifts we might have 2 separate flags:
* $shr_0$ which says that $s_0 \rightarrow s_1'$ (where $s_i$ is the $i$-th item on the stack).
* $shr_1$ which says that $s_i -> s_{i+1}'$ for $i \in 1..14$.

This way, the general shift right flag would be described as $shr = shr_0 + shr_1$. And then we could have only 15 constraints which describe right shift behavior. It does, however, make it more difficult to come up with individual shift flags. We'll basically have to map all operations resulting in right shifts into some combination of these flags. For example:

* For `dup*` operations (and probably most others) the flags would be $shr_0 = 1$ and $shr_1 = 1$.
* But for something like `u32split`, the flags would be $shr_0 = 0$ and $shr_1 = 1$.

Similar would apply to shift left operations, though, I believe there will be more variety there.

**Comment by bobbinth** [comment]
By the way, this would also apply to operations which don't shift the stack. For example, grouping the operations which don't change the stack beyond 4th element together could help us greatly reduce the number of constraints.

**Comment by grjte** [comment]
> There is one other consideration regarding left and right shifts: it might make sense to define multiple shift flags for each type of shift. This may help with constraint definitions later.
> 
> For example, for right shifts we might have 2 separate flags:
> * $shr_0$ which says that $s_0 \rightarrow s_1'$ (where $s_i$ is the $i$-th item on the stack).
> * $shr_1$ which says that $s_i -> s_{i+1}'$ for $i \in 1..14$.
> 
> This way, the general shift right flag would be described as $shr = shr_0 + shr_1 - shr_0 \cdot shr_1$. And then we could have only 15 constraints which describe right shift behavior. 

This makes sense, but looking at the list of operations that cause a right shift, `u32split` seems to be the only one that shifts the stack from an index other than 0. `u32split` is already defined by its own unique flag, since it's grouped under the range check prefix, so it seems that the rest could be grouped together and we could still do something similar to what you propose.

**Comment by grjte** [comment]
> Similar would apply to shift left operations, though, I believe there will be more variety there.

The 8 ops that I grouped together in #236 do all shift the stack from the same index (2), and these are all of the left-shift ops that shift by that amount. The two that are grouped under the range check prefix also shift the stack by the same amount (from index 3).

As you said, there is greater variety for the remaining ops:
- shift by 1: 3 ops
- shift by 3: 3 ops (2 of which use the range check prefix)
- shift by 5: 2 ops
- shift by 9: 1 op

**Comment by bobbinth** [comment]
Another approach to organizing operations could be as follows:

| bit 6 | bit 5 | bit 4 | bit 3 | bit 2 | bit 1 | bit 0 | # of ops | degree |
| :---: | :---: | :---: | :---: | :----: | :---: | :---: | :-------: | :------: |
| 0      |  x      | x      | x       | x       | x       | x      |  64          |  7         |
| 1      |  0      | 0      | x       | x       | x       | x      |  16           |  7         |
| 1      |  0      | 1      | x       | x       | x       | -       |  8           |  6         |
| 1      |  1      |  x     | x       | -       | -       | -       |  4           |  4         |

It is similar to the approach outlined in the original post, but instead of 8 degree 5 operations, we have 4 degree 4 operations. So, we get 92 total operations split as follows:

* 80 opcodes of degree 7
* 8 opcodes of degree 6
* 4 opcodes of degree 4

The assignment of operations could be as follows:
* Degree 4 ops: `PUSH`, `REPEAT`, `END`, and `HALT`.
* Degree 6 ops: `JOIN`, `SPLIT`, `LOOP`, `SPAN`, `RESPAN` + 3 more ops
* Degree 7 ops: all remaining user ops

I believe the above arrangement would still work fine for the decoder constraints (no constraint should go above 9) - but this needs to be verified. If this works out, we can reduce column count by 2 as we won't need 2 extra columns in the decoder for degree reduction.

**Comment by bobbinth** [comment]
One of the drawbacks of the above approach is that we get very few degree 6 ops available. Beyond control flow operations, we don't currently have a lot of operations which would require degree 6 flags or smaller. The ones that come to mind are `u32split` (which we actually can't put in the degree 6 bucket because it requires 4 range checks) and maybe `binacca`. But in the future we might need to have more (e.g., `call` and `syscall` will likely need to be degree 6 or smaller).

Another annoyance is that a left shift flag needs to have degree $5$ or smaller, and this means we might need to introduce an extra column for it because range-check operation group has two operations which shift the stack to the left (`u32add3` and `u32madd`), and they can't be grouped together to form a shorter than degree 6 prefix.

If we need to introduce an extra column, we could make it more generally useful than just for left shift flag. Specifically, we could arrange operations as follows:

| bit 6 | bit 5 | bit 4 | bit 3 | bit 2 | bit 1 | bit 0 | extra | # of ops | degree |
| :---: | :---: | :---: | :---: | :----: | :---: | :---: | :----: |:-------: | :------: |
| 0      |  x      | x      | x       | x       | x       | x      |  0      | 64          |  7         |
| 1      |  0      | x      | x       | x       | x       | -       |  0      | 16           |  6         |
| 1      |  1      |  x     | x       | x       | -       | -       |  1        | 8           |  4         |

In the above, $extra = b_6 \cdot b_5$ and thus, we can use for computing flags in the last row while still keeping their degree at 4.

The total number of operations we get is 88 - but I think that might be OK. With this arrangement, we could group operations as follows:

* 8 u32 operations which require range checks could go into the degree 6 bucket (e.g., with a prefix of `100`). This would also mean that if `u32add3` and `u32madd` are grouped under a common prefix, their degree would be $5$, and thus, we might be able to avoid creating an extra column for left shift flag.
* We could put most of control flow operations into the degree 4 bucket. One operation would need to be in degree 6 bucket and we can figure out which one would make the most sense to put there.
* This would leave 7 degree 6 operations + 64 degree 7 operations available for everything else.

**Comment by bobbinth** [comment]
A slight update to the above. We actually have several "high degree" user operations. These are:

* `u32split` (degree 3) - this would go into the degree 6 bucket as described above.
* `rpperm` (degree 3) - this would also go into the degree 6 bucket.
* `mpverify` (degree 3) - this would also go into the degree 6 bucket.
* `mrupdate` (degree 5) - this would need to go into the degree 4 bucket.

Thus, the overall layout changes slightly:

* 8 u32 operations (including `u32split`) which require range checks could go into the degree 6 bucket (e.g., with a prefix of 100). This would also mean that if u32add3 and u32madd are grouped under a common prefix, their degree would be , and thus, we might be able to avoid creating an extra column for left shift flag.
* `rpperm` and `mpverify` operations would also go into degree 6 bucket.
* `mrupdate` and `push` operations would go into the degree 4 bucket.
* 6 out of 8 control flow operations would go into the degree 4 bucket as well. The remaining two would need to go into the degree 6 bucket.
* This would leave 4 degree 6 operations + 64 degree 7 operations available for everything else.

**Comment by bobbinth** [comment]
Using the approach described above, I came up with the operation grouping scheme which I think satisfies all the conditions we need. The picture of how this looks like is pasted below.

A few general notes:
* We break operations into 3 high-level groups based on their maximum constraint degree:
  - 64 operations with max degree 2.
  - 16 operations with max degree 3.
  - 8 operations with max degree 5.
* The first 32 operations (prefix `00`) do not shift the stack. In this group we have 4 empty slots which we can use for future operations.
* The next 16 operations (prefix `010`) shift the stack to the left. This group is currently full, but once we remove `u32or` operation, it will have one empty slot for future operations.
* The next 16 operations (prefix `011`) shift the stack to the right. This group currently has 1 empty slot for future operations.
* The next 8 operations (prefix `100`) are u32 operations which require range checks. Having these operations here is convenient for several reasons:
  - Prefix `100` has degree 3, and thus we can support 4 range checks for these operations.
  - The two operations which result in a left shift for the stack (`u32add3` and `u32madd`) are grouped together under the common prefix `10011`. This prefix has degree 5, which is important because left shift flag cannot exceed degree 5.
  - `u32split` operation is also in this group, which is convenient because constraints for this operation have degree 3, and thus the operation flag cannot exceed degree 6.
* The next 8 operations (prefix `101`) are operations which have constraint degrees 3: `rpperm`, `mpverify` and 4 control flow operations. In this group we have 2 empty slots for future operations.
  - Again, operations requiring a left shift (`split` and `loop`) are grouped together under common prefix `10111`, thus ensuring that left shift flag has degree 5.
![image](https://user-images.githubusercontent.com/43513081/181830316-9ae75039-434d-442e-bab8-721ec1d8c02b.png)
* The last group of 8 operations contains operations with constraints of degree up to 5: `push`, `mrupdate`, and 4 control flow operations. In this group we also have 2 empty slots for future operations.
  - It is important that `end` operation has degree 4, because whether `end` operation results in a left shift is determined by multiplying the flag of `end` operation with a helper register - thus, giving the overall degree to 5 to the left shift flag.

**Comment by bobbinth** [comment]
Closed by #340
